src/controller/p2p.rs

```diff
@@ -36,90 +40,170 @@ lazy_static! {
 
-    Ok(P2PTransport {
+    let (queries_tx, queries_rx) = mpsc::channel(QUERIES_POOL_SIZE);
+    let (last_collected_log_tx, _) = watch::channel(None);
+
+    Ok(P2PController {
+        worker,
+        ping_interval,
         raw_event_stream: UseOnce::new(event_stream),
         transport_handle,
         logs_storage: LogsStorage::new(data_dir.join("logs.db").as_str()).await?,
         worker_id,
-        assignments_tx,
-        assignments_rx: UseOnce::new(assignments_rx),
         queries_tx,
         queries_rx: UseOnce::new(queries_rx),
+        last_collected_log_tx,
     })
 }
 
-impl<EventStream: Stream<Item = WorkerEvent>> P2PTransport<EventStream> {
-    async fn run_receive_loop(&self, cancellation_token: CancellationToken) {
-        let event_stream = self.raw_event_stream.take().unwrap();
-        event_stream
+impl<EventStream: Stream<Item = WorkerEvent>> P2PController<EventStream> {
+    pub async fn run(&self, cancellation_token: CancellationToken) {
+        // TODO: cancel all the tasks if one of them finishes
+        tokio::join!(
+            self.run_event_loop(cancellation_token.child_token()),
+            self.run_queries_loop(cancellation_token.child_token()),
+            self.run_ping_loop(cancellation_token.child_token(), self.ping_interval),
+            self.run_logs_loop(cancellation_token.child_token(), *LOGS_SEND_INTERVAL),
+            self.worker.run(cancellation_token.child_token()),
+        );
+    }
+
+    async fn run_queries_loop(&self, cancellation_token: CancellationToken) {
+        let queries_rx = self.queries_rx.take().unwrap();
+        ReceiverStream::new(queries_rx)
             .take_until(cancellation_token.cancelled_owned())
-            .for_each_concurrent(CONCURRENT_MESSAGES, |ev| async move {
-                match ev {
-                    WorkerEvent::Pong(pong) => self.handle_pong(pong),
-                    WorkerEvent::Query { peer_id, query } => {
-                        self.handle_query(peer_id, query).await
-                    }
-                    WorkerEvent::LogsCollected { last_seq_no } => {
-                        self.handle_logs_collected(last_seq_no).await
-                    }
-                }
+            .for_each_concurrent(CONCURRENT_QUERY_MESSAGES, |(peer_id, query)| async move {
+                self.handle_query(peer_id, query).await;
             })
             .await;
     }
 
-    async fn run_send_logs_loop(
+    async fn run_ping_loop(&self, cancellation_token: CancellationToken, ping_interval: Duration) {
+        let mut timer =
+            tokio::time::interval_at(tokio::time::Instant::now() + ping_interval, ping_interval);
+        timer.set_missed_tick_behavior(MissedTickBehavior::Delay);
+        loop {
+            tokio::select!(
+                _ = timer.tick() => {},
+                _ = cancellation_token.cancelled() => {
+                    break;
+                },
+            );
```

Comment:
Could be done without `select` by using `IntervalStream::new(timer).take_until(cancellation_token.cancelled()).for_each{|_| { .. })`
but it's mostly a matter of taste, I guess.